### DISCUSSION TITLE
How to change the default sensor mode that gets selected by the video configuration?

### DISCUSSION THREAD

**Discussion by alexandrucindea** [raspberrypi/picamera2#756] [Q&A] [answered]
I am having issues setting up my Module 3 Camera WIDE to use the full image crop. I can run .sensor_modes to see the available modes, but I can't figure out how to select one. I checked the documentation on configuring the camera, but can't really see a way to select a preconfigured mode.

**Comment by davidplowman** [comment]
Hi, you should be able to pass one of those modes directly, for example
```
config = picam2.create_preview_configuration(raw=picam2.sensor_modes[0])
```
or alternatively just list the format and size that you have identified:
```
config = picam2.create_preview_configuration(raw={'format': 'SBGGR10_CSI2P', 'size': (2304, 1296)})
```

**Reply by alexandrucindea** [reply]
Thank you for your response. While it seems to work on the preview window, i am trying to record video. When setting the size, the preview window seems to change, but when changing this in the app_recording.py example from the repository, pressing record throws a ProcessLookupError: No such process. I can't seem to figure out the reason for this behaviour. It also will not let me specify the format when creating a video config with .create_video_configuration(main={'format': 'SBGGR10_CSI2P', 'size': (2304, 1296)}).

**Reply by alexandrucindea** [reply]
`config = picam2.create_preview_configuration(raw=picam2.sensor_modes[0])`

Also, for me this throws an error: 
`ValueError: Bad key 'unoacked': valid stream configuration keys are ('format', 'size')`

**Comment by davidplowman** [comment]
Could you just confirm your Picamera2 version? There was a bug exactly like this that was fixed a little while ago. Thanks!

**Reply by alexandrucindea** [reply]
Sure, tried upgrading right now, I am on 0.3.12-2, I got the message saying that i was already on latest.

**Comment by davidplowman** [comment]
Could you include the exact and most minimal script - presumably just a handful of lines - that produce this error? Then it's easy for me to try it and know that I'm looking at the right problem. Thanks.

**Reply by alexandrucindea** [reply]
Of course. In this case I was fidgeting around with one of the examples in the repo, app_recording.py.

```
#!/usr/bin/python3

from PyQt5 import QtCore
from PyQt5.QtWidgets import (QApplication, QHBoxLayout, QLabel, QPushButton,
                             QVBoxLayout, QWidget)

from picamera2 import Picamera2
from picamera2.encoders import H264Encoder
from picamera2.outputs import FileOutput
from picamera2.previews.qt import QGlPicamera2


def post_callback(request):
    label.setText(''.join(f"{k}: {v}\n" for k, v in request.get_metadata().items()))


picam2 = Picamera2()
picam2.post_callback = post_callback
picam2.configure(picam2.create_video_configuration(main={"size": (2304, 1296)}))

app = QApplication([])


def on_button_clicked():
    global recording
    if not recording:
        encoder = H264Encoder(10000000)
        output = FileOutput("test.h264")
        picam2.start_encoder(encoder, output)
        button.setText("Stop recording")
        recording = True
    else:
        picam2.stop_encoder()
        button.setText("Start recording")
        recording = False


qpicamera2 = QGlPicamera2(picam2, width=800, height=480, keep_ar=False)
button = QPushButton("Start recording")
button.clicked.connect(on_button_clicked)
label = QLabel()
window = QWidget()
window.setWindowTitle("Qt Picamera2 App")
recording = False

label.setFixedWidth(400)
label.setAlignment(QtCore.Qt.AlignTop)
layout_h = QHBoxLayout()
layout_v = QVBoxLayout()
layout_v.addWidget(label)
layout_v.addWidget(button)
layout_h.addWidget(qpicamera2, 80)
layout_h.addLayout(layout_v, 20)
window.resize(1200, 480)
window.setLayout(layout_h)

picam2.start()
window.show()
app.exec()
```

Originally, the size in the program was set to (1280, 720). It was working fine. Now i selected this specific size because it's one of the modes listed when printing picam2.sensor_modes. It actually runs, with the preview window showing the format I desire, but when pressing record, it crashes. I just want to be able to film in the full field of view provided by the Wide module, and this is the method I found on my own. If there are any other ways, I would be really thankful if you could point them out. Thank you!

**Reply by davidplowman** [reply]
Could you post your code in a code block (a line of triple backquotes just before and just after)? That will preserve the formatting and I can try it easily. Thanks!

**Reply by alexandrucindea** [reply]
of course, sorry, i updated

**Reply by davidplowman** [reply]
Thanks.

When I run the above it fails to start the video encoder (V4L2 reports "No such process", which is a bit cryptic, to be fair). The reason is that the h.264 encode only supports up to 1080p, it won't do 2304x1296. But this seems to be a different problem from the one you reported?

When I change the configure line to
```
picam2.configure(picam2.create_video_configuration(main={"size": (800, 480)}, raw=picam2.sensor_modes[1]))
```
then it seems to work OK for me. Can you report your own findings?

**Reply by alexandrucindea** [reply]
This does seem to work, thank you. Although, it is not clear as to why. Could you help me understand this behaviour? Without specifying the raw stream with the sensor_mode[1], the format seems to remain on the first sensor format 1536x864. Why does specifying the raw with the format I desire change the recorded video? Which of the streams gets selected for recording(and why)?

**Comment by davidplowman** [comment, marked as answer]
The trouble is that when you ask for a particular output image size, libcamera has to guess which of the sensor modes to use. Within libcamera all it sees are the sensor formats and resolutions, but not the "crop" (which affects the field of view). So from its point of view, 1536x864 is the best match for 800x480. Unfortunately, this is a "cropped" mode.

But as you've seen we can override the behaviour by specifying directly what raw stream size we want. Now it will select the sensor resolution that most closely matches the raw stream size, irrespective of how much scaling happens to produce the final output images. In our case, specifying a raw stream size of 2304x1296 is effectively forcing the use of that sensor mode, which we know to be uncropped (full FoV). Hope that makes sense!

**Reply by alexandrucindea** [reply]
Definitely it does! Thank you, this does clarify it.
